lib/dry/schema/message/or.rb

```diff
@@ -1,82 +1,32 @@
 # frozen_string_literal: true
 
-require 'dry/equalizer'
+require 'dry/schema/message/or/single_path'
+require 'dry/schema/message/or/multi_path'
 
 module Dry
   module Schema
     # Message objects used by message sets
     #
     # @api public
     class Message
-      # A message sub-type used by OR operations
-      #
-      # @api public
-      class Or
-        # @api private
-        attr_reader :left
-
-        # @api private
-        attr_reader :right
-
-        # @api private
-        attr_reader :path
-
-        # @api private
-        attr_reader :_path
-
-        # @api private
-        attr_reader :messages
-
+      module Or
         # @api private
         def self.[](left, right, messages)
-          if [left, right].flatten.map(&:path).uniq.size == 1
-            new(left, right, messages)
+          msgs = [left, right].flatten
+          paths = msgs.map(&:path)
+
+          if paths.uniq.size == 1
+            SinglePath.new(left, right, messages)
           elsif right.is_a?(Array)
-            right
+            if left.is_a?(Array) && paths.uniq.size > 1
+              Or::MultiPath.new(left, right)
+            else
+              right
+            end
           else
-            [left, right].flatten.max
+            msgs.max
```

Comment:
@flash-gordon that's previous behavior that I mantain here 🤷‍♂ I honestly don't remember in which cases one of the messages must win, but it's very likely we could change this in 1.0.0 so that all messages will be included.